### ISSUE TITLE
[Proyecto | Entrega 1] Descripción de problemas

### ISSUE THREAD

**Issue by Geeermy** [IIC2343/Syllabus-2024-2#33]
# Descripción de problemas - Entrega 1

En esta _issue_, podrán encontrar un formato de ejemplo para el programa de cada problema.

**ANTES DE PARTIR:** Considere los siguientes puntos:

1. Los programas NO se evaluarán en su implementación en Vivado, solo se evaluarán a nivel de Assembly.
2. Por lo anterior, no deben preocuparse de mostrar los resultados en los _display_.
3. Al no evaluarse direccionamiento indirecto en esta entrega, no se utilizarán variables tipo _array_.
4. Los nombres de las variables y el formato del programa para _testear_ son solo de referencia, puede entregar una implementación distinta ya que la revisión será **manual**.
6. Consideraciones adicionales que deba contemplar se incluirán después del formato de cada problema.
7. Si tienen dudas adicionales, por favor realícenlas **sobre esta _issue_**.

### Formato esperado

```
DATA:
matriz_00 1 // Primera fila, primer elemento
matriz_01 2 // Primera fila, segundo elemento
matriz_10 3 // Segunda fila, primer elemento
matriz_11 4 // Segunda fila, segundo elemento
det       0 // Determinante = 1 * 4 - 2 * 3 = -2. El resultado se almacena aquí

CODE:
// Código empieza aquí.
// - SU TAREA ES COMPLETAR ESTA PARTE -
// Código termina aquí.
end:
MOV A,(det)
MOV B,(det)
JMP end

```

### Consideraciones

1. Si ejecutara este código en su placa, y estuviera correcto, el _display_ mostraría `FFFE`.
2. El rango numérico evaluado será de 0 a 255, sin considerar negativos.

### Formato esperado

```
DATA:
num 162 // Valor de entrada
cen 0   // Cantidad de centenas = 1
dec 0   // Cantidad de decenas = 6
uni 0   // Cantidad de unidades = 2

CODE:
// Código empieza aquí.
// - SU TAREA ES COMPLETAR ESTA PARTE -
// Código termina aquí. A continuación, se muestra una prueba para verificar los resultados.
end:
MOV A,(cen)
MOV B,(dec)
ADD A,B
MOV B,(uni)
ADD A,B
finish:
MOV B,A
JMP finish
```

### Consideraciones

1. Si ejecutara este código en su placa, y estuviera correcto, el _display_ mostraría `A2A2`.
2. El rango numérico evaluado será de 0 a 255, sin considerar negativos.

### Formato esperado

```
DATA:
num 384  // Valor de entrada
res 0    // Resultado: 0 = No lo es; 1 = Sí lo es. En este caso, num = 0000000110000000 (lo es).

CODE:
// Código empieza aquí.
// - SU TAREA ES COMPLETAR ESTA PARTE -
// Código termina aquí. A continuación, se muestra una prueba para verificar los resultados.
end:
MOV A,(res)
MOV B,(res)
JMP end
```

### Consideraciones

1. Si ejecutara este código en su placa, y estuviera correcto, el _display_ mostraría `0101`.
3. El rango numérico evaluado será de 0 a 65535, sin considerar negativos.
4. Se toma como palíndromo un número natural de 16 bits, es decir, números de menos bits como 5 (101) no se considerarán palíndromos.

### Formato esperado

```
DATA:
num 8 // Valor de entrada.
res 0 // Resultado = num! = 40320

CODE:
// Código empieza aquí.
// - SU TAREA ES COMPLETAR ESTA PARTE -
// Código termina aquí. A continuación, se muestra una prueba para verificar los resultados.
end:
MOV A,(res)
MOV B,A
SHR A
SHR A
SHR A
SHR A
SHR A
SHR A
SHR A
SHR A
MOV A,B
finish:
JMP finish
```

### Consideraciones

1. Si ejecutara este código en su placa, y estuviera correcto, el _display_ mostraría `9D80`.
3.  El rango numérico evaluado será de 0 a 65535, sin considerar negativos.

### Formato esperado

```
DATA:
num 339  // Valor de entrada
res 0 // Resultado = 1 - 2 + 3 - 4 + 5 - 6 + 7 - ... + 339 = -170

CODE:
// Código empieza aquí.
// - SU TAREA ES COMPLETAR ESTA PARTE -
// Código termina aquí. A continuación, se muestra una prueba para verificar los resultados.
end:
MOV A,(res)
MOV B,(res)
JMP end
```

### Consideraciones

1. Si ejecutara este código en su placa, y estuviera correcto, el _display_ mostraría `AAAA`.
2. El rango numérico evaluado será de 0 a 65000, sin considerar negativos como entrada.
4. Debe basarse en [la siguiente definición de sumatoria alternante](https://es.wikipedia.org/wiki/1_%E2%88%92_2_%2B_3_%E2%88%92_4_%2B_%E2%8B%AF).

**Comment by slawhs** [comment]
Hola! Pregunta, usar labels en la sección CODE: cuenta como direccionamiento indirecto?

**Comment by Geeermy** [comment]
@slawhs No. Entendemos por direccionamiento indirecto ejecuciones como las siguientes:

```
DATA:
  var 2

CODE:
  MOV B,var   // B = dirección de var
  MOV A,(B)   // A = 2
```

Los _labels_ a nivel de código aluden a direcciones en la memoria de instrucciones, por lo que no se relacionan con direccionamiento indirecto.

Recordatorio: No es necesario que el programa corra en su implementación de la primera entrega.

¡Saludos!

**Comment by dvillas28** [comment]
Sobre el punto 1. Entonces este ejercicio no tiene nada que ver con el Assembler? Solamente nos piden escribir Codiga en Assembly y eso es lo que van a revisar es lo que estoy entendiendo? 👀

**Comment by Geeermy** [comment]
@dvillas28 Exacto. Lo que pasa es que el Assembler de la Etapa 1 es bien limitado y tendrían que hacer muchas cosas manuales, lo que haría que sus programas fueran mucho más engorrosos. El Assembler de la Etapa 2 podrá correrlos sin problema. :)

**Comment by Geeermy** [comment]
**Nota:** Se corrigieron los valores que se debían mostrar en el _display_ en algunos problemas y, en el problema 4, se eliminó una instrucción del _label_ `end` para que lo que muestre el _display_ sea correcto (en este caso, el resultado completo del número 40320 en hexadecimal, que es `9D80`.

**Comment by fgarl** [comment]
Hola, en la página para probar el codigo assembly, al hacer operaciones con variables declaradas en data, como por ejemplo
```
DATA:

num  162 

CODE:

MOV A, (num)
MOV A, num
```


No genera ningún código de maquina, y como nuestro assembler todavía no puede leer DATA tampoco podemos probar el código ahí, que podría hacer para probar si funciona? Tendría que usar como dirección el orden en el que se van declarando las funciones y variables?

**Comment by Geeermy** [comment]
@fgarl ¿Me imagino que estás probando directo en el simulador? Puedes usar [esta herramienta](https://iic2343.ing.puc.cl/) para generar código de máquina según tus _opcodes._ :)

**Comment by fgarl** [comment]
Justo ahí lo estoy probando, y al entregarle los opcodes y el mismo código que mande arriba, no me tira nada. En cambio si hago 
`MOV A, (162)` si me genera un código de maquina.
Tambien probé con en ejemplo que se dio arriba
```
DATA:
  var 2

CODE:
  MOV B,var  
  MOV A,(B)   
```
y en ese caso se queda generando infinitamente :c

**Comment by sufigueroa** [comment]
> Justo ahí lo estoy probando, y al entregarle los opcodes y el mismo código que mande arriba, no me tira nada. En cambio si hago 
> `MOV A, (162)` si me genera un código de maquina.
> Tambien probé con en ejemplo que se dio arriba
> ```
> DATA:
>   var 2
> 
> CODE:
>   MOV B,var  
>   MOV A,(B)   
> ```
> y en ese caso se queda generando infinitamente :c


Hola!! Efectivamente, solo funciona para código tipo etapa 1. Puedo subir un fix en la tarde para que acepte labels, **pero no va a soportar instrucciones de la etapa 2**. Cuando suba el cambio avisaré por acá. Con eso deberían ser capaces de probar su código de manera correcta.

**Comment by ooscarr** [comment]
Hola, en los ejemplos nuestro código empieza en la parte `CODE:` ¿es decir que no se pueden agregar más variables adicionales en la parte `DATA:`?

**Comment by Geeermy** [comment]
> Hola, en los ejemplos nuestro código empieza en la parte `CODE:` ¿es decir que no se pueden agregar más variables adicionales en la parte `DATA:`?

@ooscarr No, es solo de referencia, pueden agregar más variables (auxiliares, por ejemplo).

¡Saludos!

**Comment by bcalfaro** [comment]
Hola! Se pueden usar subrutinas, mientras tengan las operaciones de esta etapa? Entiendo que el computador aún no las aguanta

**Comment by Geeermy** [comment]
@bcalfaro No puedes trabajar con subrutinas si no tienes las instrucciones que las implementan (CALL, RET). Ahora, si te refieres a "usarlas" con instrucciones de salto y labels, eso sí se puede hacer.
